File: code/poisson/parameter.R

# ポアソン分布 -----------------------------------------------------------------

# パラメータの可視化


# パッケージの読込 -------------------------------------------------------------

# 利用パッケージ
library(tidyverse)
library(gganimate)

# パッケージ名の省略用
library(ggplot2)


# パラメータの影響 -------------------------------------------------------------

### パラメータの設定 -----

# フレームごとのパラメータを指定
lambda_vals <- seq(from = 0, to = 10, by = 0.1)

# フレーム数を設定
frame_num <- length(lambda_vals)


### 変数の設定 -----

# x軸の範囲を設定
x_min <- 0
u <- 5
x_max <- lambda_vals |> 
  max() |> 
  (\(.) {. * 1.5})() |> # 倍率を指定
  (\(.) {ceiling(. /u)*u})() # u単位で切り上げ

# x軸の値を作成
x_vec <- seq(from = x_min, to = x_max, by = 1)


### 分布の計算 -----

# ポアソン分布の確率を計算
anime_prob_df <- tidyr::expand_grid(
  frame_i = 1:frame_num, # フレーム番号
  x       = x_vec # 確率変数
) |> # フレームごとに変数を複製
  dplyr::mutate(
    lambda = lambda_vals[frame_i], # パラメータ
    prob   = dpois(x = x, lambda = lambda) # 確率
  )


### パラメータと形状の関係 -----

# ラベル用の設定を作成
anime_label_df <- tibble::tibble(
  frame_i = 1:frame_num, 
  lambda  = lambda_vals, 
  param_lbl = paste0("lambda == ", round(lambda, digits = 2))
)

# ポアソン分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anime_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968"
  ) + # 確率
  geom_text(
    data    = anime_label_df, 
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  gganimate::transition_manual(frames = frame_i) + # フレーム制御
  scale_x_continuous(breaks = x_vec, minor_breaks = FALSE) + # x軸目盛
  theme(
    plot.subtitle = element_text(size = 50) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(clip = "off") + # (パラメータラベル用の枠外表示設定)
  labs(
    title = "Poisson Distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/poisson/parameter/parameter.mp4")
)


### パラメータと統計量の関係 -----

# 統計量を計算
anim_param_df <- tibble::tibble(
  frame_i = 1:frame_num,   # フレーム番号
  lambda  = lambda_vals,   # パラメータ
  mu      = lambda,        # 期待値
  sigma   = sqrt(lambda),  # 標準偏差
  mode    = floor(lambda), # 最頻値
  param_lbl = paste0("lambda == ", round(lambda, digits = 2)), 
  stat_lbl  = paste0(
    "mean: ", sprintf(fmt = '%.02f', domain = mu), "\n", 
    "sd:      ", sprintf(fmt = '%.02f', domain = sigma), "\n", # (スペースによる位置調整)
    "mode: ", sprintf(fmt = '%.02f', domain = mode)
  )
)

# 標準偏差の範囲を計算
anim_sd_df <- tidyr::expand_grid(
  frame_i = 1:frame_num, 
  sign    = c(-1, 1) # 符号
) |> 
  dplyr::mutate(
    lambda  = lambda_vals[frame_i], # パラメータ
    mu      = lambda,       # 期待値
    sigma   = sqrt(lambda), # 標準偏差
    label_x = mu + sign*sigma, # 描画位置
    sd_lbl = dplyr::if_else(
      condition = sign == 1, 
      true      = "+ sigma", 
      false     = "- sigma"
    )
  )


# ラベル用の文字列を作成
label_vec    <- c(
  mean = expression(E(x) == lambda), 
  sd   = expression(sqrt(V(x)) == sqrt(lambda)), 
  mode = expression(mode(x) == group(lfloor, lambda, rfloor))
)

# ポアソン分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anime_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968"
  ) + # 確率
  geom_vline(
    data    = anim_param_df, 
    mapping = aes(xintercept = mu, linetype = "mean"), 
    linewidth = 1
  ) + # 期待値の位置
  geom_vline(
    data    = anim_sd_df, 
    mapping = aes(xintercept = label_x, linetype = "sd"), 
    linewidth = 1
  ) + # 標準偏差の位置
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mu-sigma, y = -Inf, xend = mu+sigma, yend = -Inf), 
    linewidth = 1
  ) + # 標準偏差の範囲
  geom_text(
    data    = anim_sd_df,
    mapping = aes(x = label_x, y = -Inf), 
    label = "|", size = 3
  ) + # 標準偏差の指示線
  geom_vline(
    data    = anim_param_df, 
    mapping = aes(xintercept = mode, linetype = "mode"), 
    linewidth = 1
  ) + # 最頻値の位置
  geom_label(
    data    = anim_param_df, 　
    mapping = aes(x = lambda, y = Inf, label = "lambda"), 
    parse = TRUE, hjust = 0.5, vjust = 1.5, 
    fill = "gray92", label.padding = unit(0, units = "lines"), label.size = 0, 
    size = 4
  ) + # パラメータのラベル
  geom_label(
    data    = anim_param_df, 　
    mapping = aes(x = mu, y = -Inf, label = "E(x) %+-% sqrt(V(x))"), 
    parse = TRUE, hjust = 0.5, vjust = -0.2, 
    fill = "gray92", label.padding = unit(0, units = "lines"), label.size = 0, 
    size = 3
  ) + # 統計量のラベル
  geom_text(
    data    = anim_param_df, 　
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  geom_label(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = stat_lbl), 
    hjust = 0, vjust = 1, alpha = 0.5
  ) + # 統計量のラベル
  gganimate::transition_manual(frames = frame_i) + # フレーム制御
  scale_x_continuous(breaks = x_vec, minor_breaks = FALSE) + # x軸目盛
  scale_linetype_manual(
    breaks = c("mean", "sd", "mode"), 
    values = c("dashed", "dotted", "dotdash"), 
    labels = label_vec, 
    name   = "statistics"
  ) + # (凡例表示用)
  guides(
    linetype = guide_legend(override.aes = list(linewidth = 0.5))
  ) + # 凡例の体裁
  theme(
    plot.subtitle = element_text(size = 40) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(clip = "off") + # (パラメータラベル用の枠外描画設定)
  labs(
    title = "Poisson distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/poisson/parameter/stats.mp4")
)


### パラメータとモーメントの関係 -----

# ガウス分布の確率密度を計算
anime_norm_df <- tidyr::expand_grid(
  frame_i = 1:frame_num, 
  x       = seq(from = min(x_vec), to = max(x_vec), length.out = 1001)
) |> # フレームごとに複製
  dplyr::mutate(
    lambda = lambda_vals[frame_i], # パラメータ
    mu     = lambda,       # 期待値
    sigma  = sqrt(lambda), # 標準偏差
    dens   = dnorm(x = x, mean = mu, sd = sigma) # 確率密度
  )


# 統計量・モーメントを計算
anim_param_df <- tibble::tibble(
  frame_i = 1:frame_num, 
  lambda  = lambda_vals,    # パラメータ
  mu      = lambda,         # 期待値
  sigma   = sqrt(lambda),   # 標準偏差
  skew    = 1/sqrt(lambda), # 歪度
  kurt    = 1/lambda,       # 尖度
  param_lbl = paste0(
    "list(", 
    "lambda == ", round(lambda, digits = 2), ", ", 
    "mu == ", round(mu, digits = 2), ", ", 
    "sigma == ", round(sigma, digits = 2), 
    ")"
  ), 
  moment_lbl = paste0(
    "skewness: ", sprintf(fmt = '%.03f', domain = skew), "\n", 
    "kurtosis:     ", sprintf(fmt = '%.03f', domain = kurt) # (スペースによる位置調整)
  )
)

# 標準偏差の範囲用の設定を作成
anim_sd_df <- tidyr::expand_grid(
  frame_i = 1:frame_num, 
  sign    = c(-1, 1) # 符号
) |> 
  dplyr::mutate(
    lambda  = lambda_vals[frame_i], # パラメータ
    mu      = lambda,       # 期待値
    sigma   = sqrt(lambda), # 標準偏差
    label_x = mu + sign*sigma, # 描画位置
    sd_lbl = dplyr::if_else(
      condition = sign == 1, 
      true      = "+ sigma", 
      false     = "- sigma"
    )
  )


# ポアソン分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anime_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968", alpha = 0.5
  ) + # ポアソン分布の確率
  geom_point(
    data    = anime_prob_df, 
    mapping = aes(x = x, y = prob), 
    color = "#00A968", size = 3
  ) + # ポアソン分布の確率
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mu, y = 0, xend = mu, yend = Inf), 
    linewidth = 0.8, linetype = "dashed"
  ) + # 期待値の位置
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mu-sigma, y = -Inf, xend = mu+sigma, yend = -Inf), 
    linewidth = 0.8
  ) + # 標準偏差の範囲
  geom_text(
    data    = anim_sd_df,
    mapping = aes(x = label_x, y = -Inf, label = "|")
  ) + # 標準偏差の指示線
  geom_text(
    data    = anim_param_df, 
    mapping = aes(x = mu, y = -Inf, label = "mu"), 
    parse = TRUE, hjust = 0.5, vjust = -0.7, 
    size = 4
  ) + # 期待値のラベル
  geom_text(
    data    = anim_sd_df, 
    mapping = aes(x = label_x, y = -Inf, label = sd_lbl), 
    parse = TRUE, hjust = 0.5, vjust = -0.7, 
    size = 4
  ) + # 標準偏差のラベル
  geom_line(
    data    = anime_prob_df, 
    mapping = aes(x = x, y = prob, color = "pois"), 
    linewidth = 1
  ) + # ポアソン分布の確率
  geom_line(
    data    = anime_norm_df, 
    mapping = aes(x = x, y = dens, color = "norm"), 
    linewidth = 1, linetype = "dashed"
  ) + # ガウス分布の確率密度
  geom_text(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  geom_label(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = moment_lbl), 
    hjust = 0, vjust = 1, alpha = 0.5
  ) + # モーメントのラベル
  gganimate::transition_manual(frames = frame_i) + # フレーム制御
  #scale_x_continuous(breaks = x_vec, minor_breaks = FALSE) + # x軸目盛
  scale_color_manual(
    breaks = c("norm", "pois"), 
    values = c("red", "#00A968"), 
    labels = c("Gaussian", "Poisson"), 
    name   = "distribution"
  ) + # (凡例の表示用)
  guides(
    color    = guide_legend(override.aes = list(linewidth = 0.5)), 
    linetype = guide_legend(override.aes = list(linewidth = 0.5))
  ) + # 凡例の体裁
  theme(
    plot.subtitle = element_text(size = 40) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(clip = "off") + # (パラメータラベル用の枠外描画設定)
  labs(
    title = "Poisson distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability, density"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/poisson/parameter/moment.mp4")
)

File: code/poisson/parameter.py

# ポアソン分布 ------------------------------------------------------------------

# パラメータの可視化


# %%

# ライブラリの読込 ---------------------------------------------------------------

# ライブラリを読込
import numpy as np
from scipy.stats import poisson, norm
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation


# %%

# パラメータの影響 ---------------------------------------------------------------

### パラメータの設定 -----

# フレームごとのパラメータを指定
lambda_vals = np.arange(start=0.0, stop=10.1, step=0.1)

# フレーム数を設定
frame_num = len(lambda_vals)


# %%

### 変数の設定 -----

# x軸の範囲を設定
x_min = 0.0
u = 5.0
x_max = np.max(lambda_vals)
x_max *= 1.5 # 倍率を指定
x_max = np.ceil(x_max /u)*u # u単位で切り上げ

# x軸の値を作成
x_vec = np.arange(start=x_min, stop=x_max+1, step=1)


# %%

### 分布の計算 -----

# ポアソン分布の確率を計算
prob_lt = [
    poisson.pmf(k=x_vec, mu=lambda_vals[i]) for i in range(frame_num)
]


# %%

### パラメータと形状の関係 -----

# 確率軸の範囲を設定
u = 0.05
prob_max = np.max(prob_lt)
prob_max = np.ceil(prob_max /u)*u # u単位で切り上げ

# 図を初期化
fig, ax = plt.subplots(figsize=(8, 6), dpi=100, facecolor='white')
fig.suptitle('Poisson distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    lmd = lambda_vals[i]  # パラメータ
    prob_vec = prob_lt[i] # 確率
    
    # ポアソン分布を描画
    ax.bar(
        x=x_vec, height=prob_vec, 
        color='#00A968'
    ) # 確率
    ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.grid()
    ax.set_xlabel('x')
    ax.set_ylabel('probability')
    ax.set_title(f'$\\lambda = {lmd:.1f}$', loc='left')
    ax.set_ylim(ymin=0.0, ymax=prob_max) # 描画範囲を固定

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../figure/poisson/parameter/parameter.mp4', 
    progress_callback=lambda i, n: print(f'frame: {i} / {n}')
)


# %%

### パラメータと統計量の関係 -----

# 確率軸の範囲を設定
u = 0.05
prob_max = np.max(prob_lt)
prob_max = np.ceil(prob_max /u)*u # u単位で切り上げ

# 余白を追加
y_margin = 0.05
y_min = -prob_max * y_margin
y_max = prob_max * (1.0+y_margin)

# 図を初期化
fig, ax = plt.subplots(figsize=(8, 6), dpi=100, facecolor='white')
fig.suptitle('Poisson distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    lmd = lambda_vals[i]  # パラメータ
    prob_vec = prob_lt[i] # 確率

    # 統計量を計算
    mean_x = lmd           # 期待値
    sd_x   = np.sqrt(lmd)  # 標準偏差
    mode_x = np.floor(lmd) # 最頻値
    
    # ポアソン分布を描画
    ax.bar(
        x=x_vec, height=prob_vec, 
        color='#00A968'
    ) # 確率
    ax.vlines(
        x=mean_x, ymin=y_min, ymax=y_max, 
        color='black', linewidth=1.0, linestyles='--', 
        label=f'$E[x] = \\lambda = {mean_x:.2f}$'
    ) # 期待値の位置
    ax.vlines(
        x=[mean_x-sd_x, mean_x+sd_x], ymin=y_min, ymax=y_max, 
        color='black', linewidth=1.0, linestyles=':', 
        label=f'$\\sqrt{{V[x]}} = \\sqrt{{\\lambda}} = {sd_x:.2f}$'
    ) # 標準偏差の位置
    ax.hlines(
        y=0.0, xmin=mean_x-sd_x, xmax=mean_x+sd_x, 
        color='black', linewidth=1.0
    ) # 標準偏差の範囲
    ax.vlines(
        x=mode_x, ymin=y_min, ymax=y_max, 
        color='black', linewidth=1.0, linestyles='-.', 
        label=f'$mode[x] = \\lfloor \\lambda \\rfloor = {mode_x:.2f}$'
    ) # 最頻値の位置
    ax.text(
        x=mean_x, y=0.0, 
        s='$E[x] \pm \\sqrt{V[x]}$', ha='center', va='top', 
        size=8
    ) # 統計量のラベル
    ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.grid()
    ax.set_xlabel('x')
    ax.set_ylabel('probability')
    ax.set_title(f'$\\lambda = {lmd:.1f}$', loc='left')
    ax.legend(title='statistics', prop={'size': 8}, loc='upper left')
    ax.set_ylim(ymin=y_min, ymax=y_max) # (垂線との対応用)

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../figure/poisson/parameter/stats.mp4', 
    progress_callback=lambda i, n: print(f'frame: {i} / {n}')
)


# %%

### パラメータとモーメントの関係 -----

# 確率軸の範囲を設定
u = 0.05
prob_max = np.max(prob_lt)
prob_max = np.ceil(prob_max /u)*u # u単位で切り上げ

# 余白を追加
y_margin = 0.05
y_min = -prob_max * y_margin
y_max = prob_max * (1.0+y_margin)

# 図を初期化
fig, ax = plt.subplots(figsize=(8, 6), dpi=100, facecolor='white')
fig.suptitle('Poisson distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    lmd = lambda_vals[i]       # パラメータ
    pois_prob_vec = prob_lt[i] # 確率

    # 統計量を計算
    mu    = lmd          # 期待値
    sigma = np.sqrt(lmd) # 標準偏差

    # モーメントを計算
    if lmd > 0.0:
        skew = 1.0 / np.sqrt(lmd) # 歪度
        kurt = 1.0 / lmd          # 尖度
    else: # (0除算の回避用)
        skew = np.inf
        kurt = np.inf

    # ガウス分布の確率密度を計算
    norm_x_vec    = np.linspace(start=x_min, stop=x_max, num=1001)
    norm_dens_vec = norm.pdf(x=norm_x_vec, loc=mu, scale=sigma)
    
    # ラベル用の文字列を作成
    moment_str  = f'skewness: {skew:.3f} \n'
    moment_str += f'kurtosis:    {kurt:.3f}' # (スペースによる位置調整)

    # ポアソン分布を描画
    ax.bar(
        x=x_vec, height=pois_prob_vec, 
        color='#00A968', alpha=0.5
    ) # ポアソン分布の確率
    ax.scatter(
        x=x_vec, y=pois_prob_vec, 
        color='#00A968', s=30
    ) # ポアソン分布の確率
    ax.vlines(
        x=mu, ymin=y_min, ymax=y_max, 
        color='black', linewidth=1.0, linestyles='dashed'
    ) # 期待値の位置
    ax.hlines(
        y=0.0, xmin=mu-sigma, xmax=mu+sigma, 
        color='black', linewidth=1.0
    ) # 標準偏差の範囲
    for label_x, label_str in zip([mu-sigma, mu, mu+sigma], ['$-\sigma$', '$\mu$', '$+\sigma$']):
        ax.text(
            x=label_x, y=0.0, 
            s=label_str, ha='center', va='top', 
            size=10
        ) # 統計量のラベル
    ax.plot(
        x_vec, pois_prob_vec, 
        color='#00A968', linewidth=1.0, 
        label='poisson'
    ) # ポアソン分布の確率
    ax.plot(
        norm_x_vec, norm_dens_vec, 
        color='red', linewidth=1.0, linestyle='dashed', 
        label='gaussian'
    ) # ガウス分布の確率密度
    ax.text(
        x=x_min, y=prob_max, 
        s=moment_str, ha='left', va='top', 
        bbox=dict(facecolor='white', alpha=0.8, edgecolor='black', linewidth=0.5), 
        size = 10
    ) # モーメントのラベル
    ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.grid()
    ax.set_xlabel('x')
    ax.set_ylabel('probability, density')
    ax.set_title(f'$\\lambda = {lmd:.1f}, \mu = {mu:.1f}, \sigma = {sigma:.2f}$', loc='left')
    ax.legend(title='distribution', prop={'size': 8}, loc='upper right')
    ax.set_ylim(ymin=y_min, ymax=y_max) # (垂線との対応用)

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../figure/poisson/parameter/moment.mp4', 
    progress_callback=lambda i, n: print(f'frame: {i} / {n}')
)


# %%


